### PR TITLE
[Spec Decode][Perf] Replicate DSpark Markov head across TP ranks

### PR DESCRIPTION
**Pull request by mgoin**
## Summary

Replicate the DSpark Markov embedding and projection across TP ranks, removing an all-reduce and full-vocabulary gather per draft position.

## Benchmark

Qwen3-14B DSpark on B300, batch size 1, 32 input / 128 output tokens, 7 speculative tokens, dummy weights, 3 warmups and 10 measured iterations:

| TP | Sharded | Replicated | Gain |
|---:|---:|---:|---:|
| 2 | 161.5 tok/s | 166.8 tok/s | +3.3% |
| 4 | 207.0 tok/s | 214.8 tok/s | +3.8% |

## Testing

- Pre-commit hooks passed
- DSpark model registry tests: 3 passed
- Synthetic full-tensor weight loading and numerical equivalence passed

Searched open DSpark/Markov-head PRs; #47584 optimizes a separate draft LM-head path and does not overlap. AI assistance was used; the submitter reviewed the change and results.

### CODE REVIEW

File: vllm/model_executor/models/qwen3_dspark.py [hunk 1]
```diff
@@ -50,21 +54,26 @@ def __init__(
         prefix: str,
     ) -> None:
         super().__init__()
-        # TODO(ben): profile for which (if any) it makes sense to replicate or TP-shard
-        self.markov_w1 = VocabParallelEmbedding(
-            vocab_size, markov_rank, prefix=maybe_prefix(prefix, "markov_w1")
-        )
-        self.markov_w2 = ParallelLMHead(
-            draft_vocab_size, markov_rank, prefix=maybe_prefix(prefix, "markov_w2")
+        self.markov_w1 = nn.Embedding(vocab_size, markov_rank, _freeze=True)
```
Comment (benchislett) [inline vllm/model_executor/models/qwen3_dspark.py]: Why is `_freeze` needed? I assumed vllm would always run in inference mode (no gradients)?

File: vllm/model_executor/models/qwen3_dspark.py [hunk 2]
```diff
@@ -50,21 +54,26 @@ def __init__(
...
 
-    def bias(self, markov_embed: torch.Tensor, logits_processor) -> torch.Tensor:
+    def bias(
+        self,
+        markov_embed: torch.Tensor,
+        _logits_processor: LogitsProcessor,
+    ) -> torch.Tensor:
         """Vocab-size transition bias from a Markov embedding ([B, r] -> [B, V])."""
-        return logits_processor(self.markov_w2, markov_embed)
+        return self.markov_w2(markov_embed)
```
Comment (benchislett) [inline vllm/model_executor/models/qwen3_dspark.py]: Why don't we use the logits processor anymore? Are we certain it's always _only_ applying the `markov_w2(markov_embed)`?

Comment (mgoin) [reply vllm/model_executor/models/qwen3_dspark.py]: You're right, this breaks Gemma4 DSpark :(

File: vllm/model_executor/models/qwen3_dspark.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/models/qwen3_dspark.py]: This is overkill. I was just following `requires_grad=False` we do in other places

File: vllm/model_executor/layers/vocab_parallel_embedding.py [hunk 3]
```diff
@@ -323,6 +331,13 @@ def __init__(
             params_dtype=params_dtype,
             weight_loader=self.weight_loader,
         )
+        self.update_param_tp_status()
+
+    def update_param_tp_status(self):
+        for param in self.parameters():
```
Comment (benchislett) [inline vllm/model_executor/layers/vocab_parallel_embedding.py:337]: What is this for?

Comment (mgoin) [reply vllm/model_executor/layers/vocab_parallel_embedding.py:337]: This is for compat with weight reloading, see https://github.com/vllm-project/vllm/pull/48025. This is just mirroring `LinearBase` now that we have disable_tp in this class

**Review by benchislett — APPROVED** [review]
